### ISSUE TITLE
JRuby wrongly mutates hash passed in as keyword arg with indy enabled

### ISSUE THREAD

**Issue by ivoanjo** [jruby/jruby#4725]
### Environment

* JRuby: `jruby 9.1.12.0 (2.3.3) 2017-06-15 33c6439 Java HotSpot(TM) 64-Bit Server VM 25.131-b11 on 1.8.0_131-b11 [linux-x86_64]`
* Kernel: `Linux maruchan 4.11.0-041100-generic #201705041534 SMP Thu May 4 19:36:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
* Distro: `Ubuntu 17.04`

### Expected Behavior

The following script executes without issues on MRI (printing dots):

```ruby
def boom(foo:, **_); end

10.times do
  print '.'

  env = {foo: nil}.freeze
  boom(env)
end

puts
```

Result:

```bash
$ ruby trigger_bug.rb 
..........
$
```

The same example runs without issue with `--dev` or without indy:

```bash
$ jruby --dev trigger_bug.rb 
..........
$ jruby -Xcompile.invokedynamic=false trigger_bug.rb 
..........
$
```

### Actual Behavior

The same example when run under indy:

```ruby
$ jruby -Xcompile.invokedynamic=true trigger_bug.rb 
..RuntimeError: can't modify frozen Hash
                   delete at org/jruby/RubyHash.java:1572
                     boom at trigger_bug.rb:-1
  block in trigger_bug.rb at trigger_bug.rb:7
                    times at org/jruby/RubyFixnum.java:299
                   <main> at trigger_bug.rb:3
```

This bit me as the code I simplified this from was actually making use of the hash after calling the method that caused the unexpected mutation, with of course very unexpected results.

**Comment by headius** [comment]
Ok I've figured out the issue here.

When there are kwargs on the receiving end, both the interpreter and the non-indy JIT will at some point call `frobnicateKwargsArgument` which handles missing kwargs and causes duplication of hashes passed in. See for example `CompiledIRMethod.call` that receives `IRubyObject[]`.

I believe we need to move this logic into IR rather than having it done as a separate dispatch step outside of IR. However in the short term I will add this call to method bodies that need it.

**Comment by headius** [comment]
On second thought, I'm going to look into whether we can eliminate this extra step. It creates a lot of unwanted objects.

**Comment by headius** [comment]
The frobnicate mostly just peels off non-symbols from the kwargs hash so that they aren't fed into keyword argument processing. Non-symbol keys go into a separate (new) hash and are passed as an additional argument separate from the kwargs hash.

This logic should be initiated from IR rather than hidden in the interpreter or method wrapper (or in the case of the broken indy calls, within the indy binding/handles).

**Comment by headius** [comment]
I have pushed a temporary fix that basically causes indy to use a less-optimized path when the target method has keyword args, so that any incoming kwargs hash can be "frobnicated".

I also pushed in e0647809295464fb92861ef3f438bcf8aa2f993c an improvement to frobnication that should reduce the number of throw-away objects created.

More work to come but this is at least temporarily fixed on master.

**Comment by headius** [comment]
@enebo and I have been discussing with @subbuss plans to optimize hash arguments properly. The small fix will probably go into 9.1.13 and then we will look at eliminating the hash altogether (when it is not needed) in 9.2.

**Comment by enebo** [comment]
We have code starting to push through for optimizing kwargs but the basic reported problem here is fixed for the sake of 9.1.13.0 so I am resolving.

**Comment by ivoanjo** [comment]
As always, thanks for the great feedback and fix! @Talkdesk's codebases make heavy use of kwargs so looking forward to those performance improvements :)
